src/main/java/org/gradle/java/JigsawPlugin.java

```diff
@@ -0,0 +1,231 @@
+
+    private static final String JAVA_PLUGIN = "org.gradle.java";
+
+    private static final String JAVA_LIBRARY_PLUGIN = "org.gradle.java-library";
+
+    private static final String APPLICATION_PLUGIN = "application";
+
+    private static final String EXTENSION_NAME = "architecture";
+
+    private static final String LIBS_PLACEHOLDER = "APP_HOME_LIBS_PLACEHOLDER";
+
+    @Override
+    public void apply(Project project) {
+        LOGGER.debug("Applying JigsawPlugin to " + project.getName());
+        if (!project.getPlugins().hasPlugin(JAVA_PLUGIN) && !project.getPlugins().hasPlugin(JAVA_LIBRARY_PLUGIN)) {
+            project.getPluginManager().apply(JavaPlugin.class);
+        }
+        project.getExtensions().create(EXTENSION_NAME, Architecture.class);
+
+        configureJavaTasks(project);
+    }
+
+    private void configureJavaTasks(final Project project) {
+        project.afterEvaluate(new Action<Project>() {
+            @Override
+            public void execute(final Project project) {
+                configureCompileJavaTask(project);
+                configureCompileTestJavaTask(project);
+                configureJarTask(project);
+                configureTestTask(project);
+                project.getPluginManager().withPlugin(APPLICATION_PLUGIN, new Action<AppliedPlugin>() {
+                    @Override
+                    public void execute(AppliedPlugin appliedPlugin) {
+                        configureRunTask(project);
+                        configureStartScriptsTask(project);
+                    }
+                });
+            }
+        });
+    }
+
+    private void configureCompileJavaTask(final Project project) {
+        final JavaCompile compileJava = (JavaCompile) project.getTasks().findByName(JavaPlugin.COMPILE_JAVA_TASK_NAME);
+        final Architecture module = (Architecture) project.getExtensions().getByName(EXTENSION_NAME);
+        final String moduleDir =
+                compileJava.getDestinationDir().toString() + System.getProperty("path.separator") + module.getModuleName();
+        compileJava.getInputs().property("moduleName", module.getModuleName());
+        compileJava.doFirst(new Action<Task>() {
+            @Override
+            public void execute(Task task) {
+                List<String> args = new ArrayList<>();
+                args.add("--module-path");
+                args.add(compileJava.getClasspath().getAsPath());
+                args.add("-d");
+                args.add(moduleDir);
+                compileJava.getOptions().setCompilerArgs(args);
+                compileJava.setClasspath(project.files());
+            }
+        });
+        compileJava.doLast(new Action<Task>() {
+            @Override
+            public void execute(Task task) {
+                project.copy(new Action<CopySpec>() {
+                    @Override
+                    public void execute(CopySpec copySpec) {
+                        copySpec.from(moduleDir);
```

Comment:
Look no further than the java-library plugin and how we provide both a classesDir and a jar there. Similarly we could provide a "modularizedClassesDir".

> Maybe you're thinking of the --module-source-path which definitely can have multiple directories from around the filesystem contributing to a single module. But, that's at compilation time.

I was indeed.

> Also, there really isn't any reason that Gradle couldn't be rejiggered to build multiple java modules from a single Gradle project. We might not want to allow that, but we might. 

We wouldn't need several modules in one directory to support that. Could just be several directories containing one module each.

> I don't think that you actually are allowed to spread a single module over multiple directories on the --module-path.

Then I really don't understand why `--module-path` can't cope with a module without some intermediate directory. To refer back to your question from earlier:

> Suppose you construct a single directory full of multiple exploded modules. You can't put all of the module-info.class files at the top of that directory. They would overwrite one another. 

I still think this is an uncommon case and I have yet to see that anywhere else but the Jigsaw examples themselves. No build tool works like that. But if I do want to do that, then this could easily be supported. Want a single module? Just use a plain old classes dir with a module-info.java. Want multiple modules in the same folder? Just nest them and put each module-info-java in each nested folder.